types/wlr_xdg_foreign_v2.c

```diff
@@ -0,0 +1,472 @@
+#define _POSIX_C_SOURCE 200809L
+
+#include <assert.h>
+#include <string.h>
+#include <stdlib.h>
+#include <wlr/types/wlr_xdg_foreign_v2.h>
+#include <wlr/types/wlr_xdg_shell.h>
+#include <wlr/util/log.h>
+#include "util/signal.h"
+#include "xdg-foreign-unstable-v2-protocol.h"
+
+#define FOREIGN_V2_VERSION 1
+
+static const struct zxdg_exported_v2_interface xdg_exported_impl;
+static const struct zxdg_imported_v2_interface xdg_imported_impl;
+static const struct zxdg_exporter_v2_interface xdg_exporter_impl;
+static const struct zxdg_importer_v2_interface xdg_importer_impl;
+
+static struct wlr_xdg_imported_v2 *xdg_imported_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zxdg_imported_v2_interface,
+		&xdg_imported_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void xdg_imported_handle_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static bool verify_is_toplevel(struct wl_resource *client_resource,
```

Comment:
Hm, why do we need to check whether the surface is a toplevel and send a protocol error if not? The protocol doesn't define any errors like this.

Reply:
Should we instead silently fail?

Reply:
No, reading the protocol again, sending a protocol error is what we should do, but we're just missing the error values.

https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/57